File: backend/schemas/audit.py
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class AuditLogResponse(BaseModel):
    id: int
    entity_type: str
    entity_id: int
    action: str
    actor_user_id: Optional[int] = None
    actor_username: Optional[str] = None
    target_user_id: Optional[int] = None
    target_username: Optional[str] = None
    source: str
    old_data: Optional[Dict[str, Any]] = None
    new_data: Optional[Dict[str, Any]] = None
    changed_fields: Optional[Dict[str, Any]] = None
    meta: Optional[Dict[str, Any]] = None
    created_at: datetime


class AuditLogListResponse(BaseModel):
    items: List[AuditLogResponse]
    total: int
    page: int
    size: int

File: frontend/src/types/audit.ts
export interface AuditLog {
  id: number;
  entity_type: string;
  entity_id: number;
  action: 'create' | 'update' | 'delete' | string;
  actor_user_id?: number | null;
  actor_username?: string | null;
  target_user_id?: number | null;
  target_username?: string | null;
  source: string;
  old_data?: Record<string, unknown> | null;
  new_data?: Record<string, unknown> | null;
  changed_fields?: Record<string, { old?: unknown; new?: unknown }> | null;
  meta?: Record<string, unknown> | null;
  created_at: string;
}

export interface AuditLogListResponse {
  items: AuditLog[];
  total: number;
  page: number;
  size: number;
}

export interface AuditLogQueryParams {
  entity_type?: string;
  entity_id?: number;
  action?: string;
  page?: number;
  size?: number;
}
